File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/CreateNewBurracoGameCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.BurracoGame
import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.waitingPlayers.BurracoGameWaitingPlayers
import com.abaddon83.cardsGames.burracoGames.ports.GameRepositoryPort
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.Future

case class CreateNewBurracoGameCmd(
                                 gameIdentity: GameIdentity
                               ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}


class CreateNewBurracoGameHandler(
                            gameRepositoryPort: GameRepositoryPort
                          )
                          (implicit val ec: scala.concurrent.ExecutionContext) extends CommandHandler[CreateNewBurracoGameCmd] {

  override def handleAsync(command: CreateNewBurracoGameCmd): Future[Unit] = {
    assert(!gameRepositoryPort.exists(command.gameIdentity), s"GameIdentity ${command.gameIdentity} already exist")
    Future {
      createNewBurracoGame(command.gameIdentity)
    }
  }

  override def handle(command: CreateNewBurracoGameCmd): Unit = {
    assert(!gameRepositoryPort.exists(command.gameIdentity), s"GameIdentity ${command.gameIdentity} already exist")
    createNewBurracoGame(command.gameIdentity)
  }

  private def createNewBurracoGame(gameIdentity: GameIdentity): BurracoGameWaitingPlayers = {
    val burracoGameWaitingPlayer = BurracoGame.createNewBurracoGame(gameIdentity)
    gameRepositoryPort.save(burracoGameWaitingPlayer)
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commands/CreateNewBurracoGameCmd.kt
package com.abaddon83.burracoGame.commands

import com.abaddon83.burracoGame.adapters.burracoGameRepositoryAdapters.inMemories.BurracoGameDB
import com.abaddon83.burracoGame.domainModels.BurracoGame
import com.abaddon83.burracoGame.ports.BurracoGameRepositoryPort
import com.abaddon83.burracoGame.shared.games.GameIdentity
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import com.abaddon83.utils.logs.WithLog
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*

data class CreateNewBurracoGameCmd(
        val gameIdentity: GameIdentity,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp):Command {

    constructor(gameIdentity: GameIdentity): this(
            gameIdentity = gameIdentity,
            commandId = UUID.randomUUID(),
            commandTimeStamp = Timestamp(System.currentTimeMillis()))

}

class CreateNewBurracoGameHandler() : CommandHandler<CreateNewBurracoGameCmd>, KoinComponent, WithLog() {

    private val repository: BurracoGameRepositoryPort by inject()

    override fun handle(command: CreateNewBurracoGameCmd) {
        createBurracoGame(command)
    }

    override suspend fun handleAsync(command: CreateNewBurracoGameCmd) {
        createBurracoGame(command)
    }

    private fun createBurracoGame(command: CreateNewBurracoGameCmd){
        check(!repository.exists(command.gameIdentity)) {
            log.warn("${command.gameIdentity} already exist")
            "${command.gameIdentity} already exist"
        }
        val burracoGameWaitingPlayer = BurracoGame.create(command.gameIdentity)
        repository.save(burracoGameWaitingPlayer)
        log.info("${command.gameIdentity} created")
    }
}
